hazelcast/src/main/java/com/hazelcast/query/SqlPredicate.java

```diff
@@ -317,7 +317,7 @@ private void readObject(java.io.ObjectInputStream in)
         Predicate[] predicates;
         if (klass.isInstance(predicateLeft)) {
             subpredicatesLeft = ((CompoundPredicate) predicateLeft).getPredicates();
-            if (predicateRight instanceof CompoundPredicate) {
+            if (predicateRight.getClass() == klass) {
```

Comment:
I'm not familiar with the problem, so I don't understand the solution. Could you shine some light on it so I can give my thumbs up?